### DISCUSSION TITLE
Configuration phase 2 (Navigation drawer configuration implementation)

### DISCUSSION THREAD

**Discussion by dubdabasoduba** [opensrp/fhircore#1304] [Ideas]
What is the vision of the configuration?
---
- We would like to achieve a configurable register flow that starts from the **Navigation drawer** to the **Client profile**.
- This will enable us to configure the following 
    - The menu item for a register and when it's displayed on the navigation drawer. 
    - The register that opens after the menu above is clicked, the configuration of the data to be displayed on the register. 
    - The profile that is opened after the register row is clicked, should also define what is displayed on the profile and the menu actions to be performed on the profile. 

Navigation Drawer configuration
---
**What we want to achieve**
- To create a configuration file to define what menus are shown on the navigation drawer. 
- We need to identify the following
    - The menus must always be shown. 
    - Menus that can be turned off and on. 
    - What registers can be shown on the navigation drawer. Here we need to identify the following
        - The main register. This will be the first register on the register list.
        - What other registers will be shown on the navigation drawer as menus. 
        - What registers show on the _"other client"_ registers menu item. Here we define the registers to be shown on the bottom sheets. 

**How to achieve the configuration**
- This configuration is using the following [mock up](https://www.figma.com/file/gDW2MxoHNI77hv1O49xgAk/FHIR-Apps-~-Design-System?node-id=10543%3A110290) as the source of truth for the Navigation menu UI. 

1. Menu identification
     - Static Menus/UI **(Will always be shown and have the same functionality)**
         - Top Bar (App Name & version)
         - Bottom Section (Sync & Sync timestamp)
         - Settings/Profile 
     - Semi Static Menus **(MIght be hidden and can be slightly configured)**
         - Main registration button (Add Family button)
         - Reports
         - Device to Device data transfer
         - Other Patients/Clients' menu
            - This opens the bottom sheet shown [here](https://www.figma.com/file/gDW2MxoHNI77hv1O49xgAk/FHIR-Apps-~-Design-System?node-id=10895%3A95946).  
     - Dynamic Menus **(Will always change)**
         - Registers menus
2. Configuration of the different menu types. 
     - **Static menus/UI** 
         - These UI pieces will always be shown to the users. The data shown on them will either come from the app config or a calculation set on the app. 
     - **Semi Static Menus**
         - Main Registration (Add family button)
             - This should be hidden or shown based on a configuration. If it is set to be shown then the last next 2 list items should always be available. 
             - The naming should also be configurable
             - The Questionnaire it opens Should also be configurable.
         - Reports
             - This should be hidden or shown based on a configuration.
             - The naming should also be configurable
         - Device to Device data transfer
             - This should be hidden or shown based on a configuration.
             - The naming should also be configurable
         - Other Patients/Clients' menu
             - This should be hidden or shown based on a configuration
             - The naming should be configurable
             - The menus to be shown on the bottom sheet should be as configurable as the Dynamic menus below. 
     - **Dynamic Menus**
         - These menus are used to define the different registers the users get access to. The following things have to be configurable. 
             - Naming of each register menu. 
             - Icons to the used. 
             - Define whether we need to show or hide the counts against each of the menus. 
             - Each menu should have a unique id assigned to it. 
             - The menus will also have an `onClick` attribute. 
                 - This attribute will be used to identify the register `config` to load when the button is clicked. The config defines the register during runtime. 

**Configuration JSON proposal**

``` json
{
  "appId": "ecbis-saa",
  "classification": "Navigation",
  "staticMenu": [
    {
      "id": "main_add_button",
      "visible": true,
      "display": "Add New Household",
      "actions": [
        {
          "type": "onClose",
          "workflow": "launch_profile",
          "id": "patient_profile"
        }
      ],
      "questionnaire": {
        "id": "82952",
        "title": "Add Family",
        "saveButtonText": "ADD FAMILY",
        "setPractitionerDetails": true,
        "setOrganizationDetails": true
      }
    },
    {
      "id": "reports",
      "visible": true,
      "display": "Reports"
    },
    {
      "id": "p2p_sync",
      "visible": true,
      "display": "Transfer Data"
    },
    {
      "id": "profile",
      "visible": true,
      "display": "Profile"
    }
  ],
  "clientRegister": [
    {
      "id": "all_households",
      "display": "All Households",
      "icon": "",
      "showCount": true,
      "actions": [
        {
          "type": "onClick",
          "workflow": "launch_register",
          "id": "all_household_register"
        },
        {
          "type": "count",
          "rules": [
            {
              "name": "householdCount",
              "condition": "",
              "action": [
                "householdCountValue = helper.count('all_household_register')"
              ]
            }
          ]
        }
      ]
    },
    {
      "id": "child",
      "display": "Child",
      "icon": "",
      "showCount": true,
      "onClick": {
        "register": "register_id"
      }
    },
    {
      "id": "sick",
      "display": "Sick",
      "icon": "",
      "showCount": true,
      "onClick": {
        "register": "register_id"
      }
    },
    {
      "id": "priority",
      "display": "Today's Priority",
      "icon": "",
      "showCount": true,
      "onClick": {
        "register": "register_id"
      }
    }
  ],
  "otherClients": {
    "visible": true,
    "display": "Other Patients",
    "bottomSheetRegister": [
      {
        "id": "bottom_all_households",
        "display": "All Households",
        "showCount": true,
        "onClick": {
          "register": "register_id"
        }
      },
      {
        "id": "bottom_child",
        "display": "Child",
        "showCount": true,
        "onClick": {
          "register": "register_id"
        }
      },
      {
        "id": "bottom_sick",
        "display": "Sick",
        "showCount": true,
        "onClick": {
          "register": "register_id"
        }
      },
      {
        "id": "bottom_priority",
        "display": "Today's Priority",
        "showCount": true,
        "onClick": {
          "register": "register_id"
        }
      }
    ]
  }
}
```
Next steps --> [Configuration phase 2 (Registers configuration implementation) ](https://github.com/opensrp/fhircore/discussions/1340)

**Comment by pld** [comment]
Looks great, for Register config do we have registers where the columns and the contents of the rows for each column vary? Then we'd want to add that too

**Reply by FikriMilano** [reply]
Do you mean between each register, the row content/design may vary?

**Reply by pld** [reply]
Yes correct

**Reply by FikriMilano** [reply]
We do have something called `patient_register_row` config for that. 
It's just that I'm not sure if it's still compatible with the current Single APK Approach.

Anyway, we can note that down.

**Reply by dubdabasoduba** [reply]
@FikriMilano Do you have a link to this?

**Reply by pld** [reply]
OK cool, maybe we'd want to migrate that over

**Reply by FikriMilano** [reply]
@dubdabasoduba 
`patient_register_row`
https://fhir.labs.smartregister.org/fhir/Binary/63002

**Comment by FikriMilano** [comment]
@dubdabasoduba
I have question, quoting this:

> 1. Menu identification
>     - Static Menus/UI **(Will always be shown and have the same functionality)**
>         - Top Bar (App Name & version)
>         - Bottom Section (Sync & Sync timestamp)
>         - Settings/Profile 

At `Settings/Profile`
By `Profile` do you mean it will open the logged in user's profile page? that will contain settings e.g. logout button, etc

**Reply by dubdabasoduba** [reply]
> @dubdabasoduba I have question, quoting this:
> 
> > 1. Menu identification
> >    
> >    * Static Menus/UI **(Will always be shown and have the same functionality)**
> >      
> >      * Top Bar (App Name & version)
> >      * Bottom Section (Sync & Sync timestamp)
> >      * Settings/Profile
> 
> At `Settings/Profile` By `Profile` do you mean it will open the logged in user's profile page? that will contain settings e.g. logout button, etc

@FikriMilano Yes.

**Reply by FikriMilano** [reply]
Got it

**Comment by FikriMilano** [comment]
@dubdabasoduba 
I saw in the designs, there are some examples that have different menu item titles, but with the same functionality.

For comparison:
- https://www.figma.com/file/gDW2MxoHNI77hv1O49xgAk/FHIR-Apps-~-Design-System?node-id=10543%3A110303
- https://www.figma.com/file/gDW2MxoHNI77hv1O49xgAk/FHIR-Apps-~-Design-System?node-id=10543%3A110316

Can I assume that we want to make the menu item titles configurable?

**Reply by dubdabasoduba** [reply]
@FikriMilano Yes that's the aim fo the Navigation Drawer configurations settings.

**Reply by FikriMilano** [reply]
Got it

**Comment by ellykits** [comment]
Can we use `camelCase` for the JSON properties for the configs.

**Reply by ellykits** [reply]
This should be consistent with the naming convention used for FHIR resource JSON properties.

**Comment by FikriMilano** [comment]
Could we remove the `priority` attribute? since it's essentially the same as using the original ordering

**Reply by pld** [reply]
can you explain that a bit?

**Reply by dubdabasoduba** [reply]
Looking at the `clientRegister` attribute, we have JSON objects in the array. What @FikriMilano is suggesting is removing the `priority` atribute in each object and using the objects position in the array to define it's position on the register menu list on the navigation drawer.

**Reply by pld** [reply]
OK I see, I think it's ok to remove. If having it as a `priority` makes the code smaller or clearer, let's keep the attribute, but if it's neutral we can remove.

It's "hidden information" or a "convention" if we don't have it the JSON, so that's not great, but I don't think that's a strong enough argument to keep it in since we're still figuring out what the configuration is

**Reply by dubdabasoduba** [reply]
@pld I think it's a strong enough argument to have it clearly highlighted on the documentation.

**Comment by ellykits** [comment]
How will we configure this type of navigation (screen to screen)? example is in Quest/G6PD workflows where we use different Details screens.  

![app_navigation](https://user-images.githubusercontent.com/26865532/172408797-e69aeacc-ffac-4226-b938-b4a6a9b6b03b.jpg)

**Reply by pld** [reply]
Can you explain a bit more, how is this relevant to the menu configuration?

**Reply by ellykits** [reply]
This is not related to the menu configuration but application screen to screen navigation. I added it here as part of navigation discussion. We can move it to a separate GitHub discussion.

**Reply by ellykits** [reply]
To explain it a bit more consider this quest/g6pd workflow:
1. Login -> go to Register -> navigate to Profile -> On click questionnaire responses/test results on profile, navigate to the default details screen when using quest as appId otherwise go to a custom screen that was implemented to display g6pd Results.

Above is just one scenario where we were switching screens based on the appId. Question is what plan do we have to configure app screen routes?

**Reply by pld** [reply]
Gotcha cool, yea so what do we want to allow for the configurability of sub-screens from within a profile, cool, yea can you start a new discussion about this? We'll need to address it, but I don't think we have a need for it w/current active projects

**Reply by ellykits** [reply]
This may as well be related to menu configurations. Consider a situation we would want to use same action to switch between different screens depending on the workflow. For example assume we have 2 types of reporting screens and we would want to switch between these two depending on the workflow, when the user clicks the Report menu action on the drawer layout.

**Reply by ellykits** [reply]
For the registers I think the question would be how to do we configure the menu actions

**Reply by pld** [reply]
K, not sure I'm following, so in the menu you would like to be able to select either of those 2 reporting screens?

**Reply by pld** [reply]
OK thanks @ellykits I think I understand this better now. It sounds like what we need to solve for here is that menu content may be tied to the workflow, so you might need to specify that in order to be able to render the appropriate menu content.

If we have a defined set of workflow names, we could have another level like
```
profile
  eir
     [menu items...]
  anc
     [menu items...]
```
then when defining your profile menu you do it for a specific workflow

**Reply by ellykits** [reply]
@dubdabasoduba What are your thoughts?

**Reply by pld** [reply]
Also, I don't think we need to solve this now (or do we? is this something we have in the app right now?), and I do think the approach we're taking can be extended to handle this

**Reply by dubdabasoduba** [reply]
@ellykits @pld I am looking at the problem in from 2 catergories
1. What do you open on click
2. what do you navigate to on close. 

The immediate problem we have to solve for is on the main registration button. 
1. We need to solve for it closing to the patient profile or to the register page. 
     - This close happens after the registration questionnaire is closed by saving the patient. 
     - If the questionnaire is closed by any other occurance then we automatically move back to the register page. 

- For this registration button situation I would propose having an onClose navigation workflow point in the app. 
- This point would have an ID in order to indicate where it is triggered.  This point and Id are then hardcode in the app. When closing a questionnaire we would first check whether the trigger menu item had an onClose config. If yes then check it the config Id matches the workflow point. 
    - If it matches then we redirect to the configuration defination
    - If not then we use the code define default. In the registration case moving back the register is the default. 

- The JSON on the button will be updated to look like the sample below.
```json
{
  "id": "main_add_button",
  "visible": true,
  "name": "Add New Household",
  "onClose": {
    "id": "registration_redirect",
    "workflow": "redirect_to_profile"
  }
}
``` 

- As for the ones on the profile I think we need to identify the diferent flows we can have then try to solve for them using the 2 categories mentioned above (if possible)
